### PR TITLE
Add unit tests for template optimization helpers

### PR DESCRIPTION
**Pull request by tenpercent**
## Summary

Add Google Tests for the optimized template utilities from the PR stack to ensure correctness and prevent regressions.

## Tests Added

| Test File | Tests | Coverage |
|-----------|-------|----------|
| `unit_sequence.cpp` | 4 new | `sequence_gen` with custom functors |
| `unit_sequence_helper.cpp` | 18 new | `generate_identity_sequences`, `find_in_tuple_of_sequences`, `sequence_find_value`, `unpack_and_merge_sequences` |
| `unit_container_helper.cpp` | 12 new | `container_concat`, `make_uniform_tuple` |
| `unit_tensor_descriptor_helper.cpp` | 18 new | `make_naive_tensor_descriptor`, `make_naive_tensor_descriptor_packed`, `make_naive_tensor_descriptor_aligned` |

**Total: 52 new tests**

## Test Coverage by PR

- **#3585**: `sequence_gen` with `__make_integer_seq`
- **#3588**: `generate_identity_sequences` helper
- **#3590**: `container_concat`, `make_uniform_tuple`
- **#3596**: `make_naive_tensor_descriptor` public API
- **#3600**: `find_in_tuple_of_sequences`, `sequence_find_value`, `unpack_and_merge_sequences`

## PR Stack

| # | PR | Description |
|---|-----|-------------|
| 1 | #3585 | sequence_gen with `__make_integer_seq` |
| 2 | #3588 | generate_identity_sequences helper |
| 3 | #3589 | Named functors in transform_tensor_descriptor |
| 4 | #3590 | container_concat optimization |
| 5 | #3596 | O(1) pack expansion rewrites |
| 6 | #3600 | TensorDescriptor/TensorAdaptor lambda elimination |
| 7 | #3608 | Build time optimization documentation |
| 8 | **This PR** | **Unit tests for optimization helpers** |

Tracking issue: #3575

### CODE REVIEW

File: test/util/unit_tensor_descriptor_helper.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+// Test compute_element_space_size helper directly
+// Formula: 1 + sum((length[i] - 1) * stride[i])
+TEST(ComputeElementSpaceSize, Simple2D)
+{
+    // 4x4 tensor with row-major strides [4, 1]
+    // element_space_size = 1 + (4-1)*4 + (4-1)*1 = 1 + 12 + 3 = 16
+    constexpr auto lengths = make_tuple(Number<4>{}, Number<4>{});
+    constexpr auto strides = make_tuple(Number<4>{}, Number<1>{});
+    constexpr auto result  = detail::compute_element_space_size(lengths, strides, Sequence<0, 1>{});
```
Comment (shumway) [inline test/util/unit_tensor_descriptor_helper.cpp]: Why are we testing an implementation detail?

File: test/util/unit_tensor_descriptor_helper.cpp [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+{
+    // 8x5 tensor with broadcast on first dimension (stride 0)
+    // element_space_size = 1 + (8-1)*0 + (5-1)*1 = 1 + 0 + 4 = 5
+    constexpr auto lengths = make_tuple(Number<8>{}, Number<5>{});
+    constexpr auto strides = make_tuple(Number<0>{}, Number<1>{});
+    constexpr auto result  = detail::compute_element_space_size(lengths, strides, Sequence<0, 1>{});
+    EXPECT_EQ(result, 5);
+}
+
+// Test make_naive_tensor_descriptor uses compute_element_space_size correctly
```
Comment (shumway) [inline test/util/unit_tensor_descriptor_helper.cpp]: This is the right way to test the element space size.

File: test/util/unit_tensor_descriptor_helper.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+TEST(MakeNaiveTensorDescriptor, ElementSpaceSize2D)
+{
+    constexpr auto lengths = make_tuple(Number<4>{}, Number<4>{});
```
Comment (shumway) [inline test/util/unit_tensor_descriptor_helper.cpp]: Can we avoid duplicate values unless we need to?  Maybe use lengths 5, 4 with strides 4, 1?

File: test/util/unit_tensor_descriptor_helper.cpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    constexpr auto result =
+        detail::compute_element_space_size(lengths, strides, Sequence<0, 1, 2>{});
+    EXPECT_EQ(result, 24);
+}
+
+TEST(ComputeElementSpaceSize, WithPadding)
+{
+    // 4x4 tensor with padded strides [8, 1] (8 elements per row instead of 4)
+    // element_space_size = 1 + (4-1)*8 + (4-1)*1 = 1 + 24 + 3 = 28
+    constexpr auto lengths = make_tuple(Number<4>{}, Number<4>{});
```
Comment (shumway) [inline test/util/unit_tensor_descriptor_helper.cpp]: With padding we can make all the different (without common factors, maybe prime numbers), e.g., lengths 11, 7, 3 strides 2, 97, 23,
  
This way the integer math is more likely to catch bad index.
This example is a valid stride, but doesn't have a simple meaning (some tiling ops might not work, but it's a good edge case for testing the element space size).

File: test/util/unit_tensor_descriptor_helper.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+{
+    // 8x5 tensor with broadcast on first dimension (stride 0)
+    // element_space_size = 1 + (8-1)*0 + (5-1)*1 = 5
+    constexpr auto lengths = make_tuple(Number<8>{}, Number<5>{});
+    constexpr auto strides = make_tuple(Number<0>{}, Number<1>{});
+    constexpr auto desc    = make_naive_tensor_descriptor(lengths, strides);
+    EXPECT_EQ(desc.GetElementSpaceSize(), 5);
+}
+
+// Test with runtime values (index_t instead of Number<>)
```
Comment (shumway) [inline test/util/unit_tensor_descriptor_helper.cpp:101]: These are good runtime value tests, but again let's avoid square shapes.

Comment (shumway) [reply test/util/unit_tensor_descriptor_helper.cpp:101]: Also don't test implementation details.

**Review by shumway — COMMENTED** [review]
This looks good, thanks! It looks like CI failed, so we'll have to recheck. My only feedback is for the element space size:

1.Don't test implementation details (avoid namespace detail::)
2. Avoid square shapes, or at least favor rectangular cases.
3. Prime number or relatives primes can help catch more edge cases (avoid accidentally getting the right answer from a bad example

(using primes or relative primes can lead to big numbers, so you may want a simpler test case, too)

File: test/util/unit_tensor_descriptor_helper.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+{
+    // 4x4 tensor with padded strides [8, 1] (8 elements per row instead of 4)
+    // element_space_size = 1 + (4-1)*8 + (4-1)*1 = 1 + 24 + 3 = 28
+    constexpr auto lengths = make_tuple(Number<4>{}, Number<4>{});
+    constexpr auto strides = make_tuple(Number<8>{}, Number<1>{});
+    constexpr auto result  = detail::compute_element_space_size(lengths, strides, Sequence<0, 1>{});
+    EXPECT_EQ(result, 28);
+}
+
+TEST(ComputeElementSpaceSize, ColumnMajor)
```
Comment (shumway) [inline test/util/unit_tensor_descriptor_helper.cpp]: This square gives the same answer for column major or row major, so it's not really testing column major. For packed arrays, ordering doesn't change the packed size.

File: test/util/unit_tensor_descriptor_helper.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+// Test with runtime values (index_t instead of Number<>)
+TEST(ComputeElementSpaceSizeRuntime, Simple2D)
+{
+    const auto lengths = make_tuple(index_t{4}, index_t{4});
+    const auto strides = make_tuple(index_t{4}, index_t{1});
+    const auto result  = detail::compute_element_space_size(lengths, strides, Sequence<0, 1>{});
+    EXPECT_EQ(result, 16);
+}
+
+TEST(ComputeElementSpaceSizeRuntime, WithPadding)
```
Comment (shumway) [inline test/util/unit_tensor_descriptor_helper.cpp]: There are two kinds of padding:

1. Padding where there is just space for extra elements. That is, you could add more elements without changing strides and get back to a packed array. These paddings often come up taking a stride slice or region out of a packed tensor. (example: shape 2,2,2 strides 1, 3, 9, which is like part of a 3,3,3 tensor)

2. Padding where there are funny offsets that don't correspond to any packed array. Those are probably worth testing just to verify there aren't any constraints on the implementation. (A simple example: shape 2,2,2 strides 1,3,7, so that elements are at offsets 0, 1, 3, 4, 7, 8, 10, 11)

File: test/util/unit_tensor_descriptor_helper.cpp (hunk 6 above)
Comment (tenpercent) [reply test/util/unit_tensor_descriptor_helper.cpp]: I think it's still worth testing different patterns of input strides

**Review by shumway — APPROVED** [review]
Thanks, looks good!

**Comment by tenpercent** [comment]
Closing this PR as all unit tests have been distributed to their respective PRs:

- 4 tests for sequence_gen → PR #3585
- 7 tests for generate_identity_sequences & unpack_and_merge_sequences → PR #3628 (merged #3588+#3589)
- 12 tests for container_concat & make_uniform_tuple → PR #3590
- 18 tests for make_naive_tensor_descriptor → PR #3596
- 11 tests for sequence_find_value & find_in_tuple_of_sequences → PR #3600

All PRs now include their own tests and target develop independently.
